xlators/cluster/afr/src/afr-common.c

```diff
@@ -5097,18 +5098,134 @@ afr_lk_unlock(call_frame_t *frame, xlator_t *this)
+                                     afr_private_t *priv, afr_local_t *local)
+{
+    dict_t *lockinfo = NULL;
+    int ret = -1;
+    int index = 0;
+
+    for (index = 0; index < priv->child_count; index++) {
+        if (fd_ctx->opened_on[index] == AFR_FD_OPENED &&
+            priv->child_up[index] && !local->need_migrate_lock[index]) {
+            ret = syncop_fgetxattr(priv->children[index], fd, &lockinfo,
+                                   GF_XATTR_LOCKINFO_KEY, NULL, NULL);
+            if (ret == 0) {
+                break;
+            }
+        }
+    }
+
+    if (ret) {
+        goto out;
+    }
+
+    if (!dict_get(lockinfo, GF_XATTR_LOCKINFO_KEY)) {
+        dict_unref(lockinfo);
+        lockinfo = NULL;
+        goto out;
+    }
+
+out:
+    return lockinfo;
+}
+
+static int
+afr_migrate_locks(afr_private_t *priv, afr_local_t *local)
+{
+    int ret = 0;
+    int index = -1;
+    fd_t *fd = NULL;
+    dict_t *lockinfo = NULL;
+    afr_fd_ctx_t *fd_ctx = NULL;
+    xlator_t *this = THIS;
+
+    if (local == NULL) {
+        ret = -1;
+        goto out;
+    }
+
+    fd = local->fd;
+    if (!fd->lk_ctx || !afr_is_need_migrate_lock(local, priv->child_count)) {
+        ret = 0;
+        goto out;
+    }
+
+    fd_ctx = afr_fd_ctx_get(fd, this);
+    if (!fd_ctx) {
+        ret = 0;
+        goto out;
+    }
+
+    lockinfo = afr_get_lock_info_from_healthy_brick(fd_ctx, fd, priv, local);
+    if (lockinfo == NULL) {
+        ret = -1;
+        goto out;
+    }
+
+    local->loc.inode = inode_ref(fd->inode);
+    for (index = 0; index < priv->child_count; index++) {
+        if (!local->need_migrate_lock[index])
+            continue;
+        if (!priv->child_up[index]) {
+            continue;
+        }
+
+        if (fd_ctx->opened_on[index] == AFR_FD_NOT_OPENED) {
+            if (IA_IFDIR == fd->inode->ia_type) {
+                syncop_opendir(priv->children[index], &local->loc, fd, NULL,
+                               NULL);
+            } else {
+                syncop_open(priv->children[index], &local->loc,
+                            fd_ctx->flags & ~(O_CREAT | O_EXCL | O_TRUNC), fd,
+                            NULL, NULL);
+            }
+        }
+
+        ret = syncop_fsetxattr(priv->children[index], fd, lockinfo, 0, NULL,
+                               NULL);
+        if (ret) {
+            goto out;
+        }
+        local->need_migrate_lock[index] = 0;
+    }
+
+out:
+    if (lockinfo != NULL)
+        dict_unref(lockinfo);
+    return ret;
+}
+
 int32_t
 afr_lk_cbk(call_frame_t *frame, void *cookie, xlator_t *this, int32_t op_ret,
            int32_t op_errno, struct gf_flock *lock, dict_t *xdata)
 {
     afr_local_t *local = NULL;
     afr_private_t *priv = NULL;
     int child_index = -1;
+    int ret = -1;
 
     local = frame->local;
     priv = this->private;
 
     child_index = (long)cookie;
+    local->need_migrate_lock[child_index] = 1;
```

Comment:
This is dangerous. This update is done without locks, which means that this '1' could be seen by another lk request that is running `afr_migrate_locks()`. This will make the other request to try to fix locks on a brick that is already healthy.